File: ios/Sources/MqttQuicPlugin/MqttQuicPlugin.swift
//
// MqttQuicPlugin.swift
// MqttQuicPlugin
//
// Capacitor plugin bridge. Phase 3: connect/publish/subscribe call into MQTTClient.
//

import Foundation
import Capacitor

@objc(MqttQuicPlugin)
public class MqttQuicPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "MqttQuicPlugin"
    public let jsName = "MqttQuic"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "ping", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendKeepalive", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "publish", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "subscribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsubscribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testHarness", returnType: CAPPluginReturnPromise)
    ]

    private var client = MQTTClient(protocolVersion: .auto)

    @objc override public func load() {}

    @objc func sendKeepalive(_ call: CAPPluginCall) {
        let timeoutMs = UInt64(call.getInt("timeoutMs") ?? 5000)
        let clamped = min(max(timeoutMs, 1000), 15000)
        Task {
            do {
                let ok = await client.sendMqttPing(timeoutMs: clamped)
                DispatchQueue.main.async { call.resolve(["ok": ok]) }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }

    @objc func ping(_ call: CAPPluginCall) {
        let host = call.getString("host") ?? ""
        let port = call.getInt("port") ?? 1884
        guard !host.isEmpty else {
            call.reject("host is required")
            return
        }
        if NGTCP2Client.ping(host: host, port: UInt16(port)) {
            call.resolve(["ok": true])
        } else {
            call.reject("Server unreachable (UDP ping to \(host):\(port) failed)")
        }
    }

    private func bundledCaPath() -> String? {
        // Prefer app bundle so TLS works when installed from npm pack (pod resources may be in separate .bundle)
        if let mainPath = Bundle.main.path(forResource: "mqttquic_ca", ofType: "pem"),
           let mainContents = try? String(contentsOfFile: mainPath),
           mainContents.contains("BEGIN CERTIFICATE") {
            return mainPath
        }
        let bundle = Bundle(for: MqttQuicPlugin.self)
        guard let path = bundle.path(forResource: "mqttquic_ca", ofType: "pem") else {
            return nil
        }
        if let contents = try? String(contentsOfFile: path),
           contents.contains("BEGIN CERTIFICATE") {
            return path
        }
        return nil
    }

    @objc func connect(_ call: CAPPluginCall) {
        let host = call.getString("host") ?? ""
        let port = call.getInt("port") ?? 1884
        let clientId = call.getString("clientId") ?? ""
        let username = call.getString("username")
        let password = call.getString("password")
        let cleanSession = call.getBool("cleanSession") ?? true
        let keepalive = call.getInt("keepalive") ?? 20
        let protocolVersionStr = call.getString("protocolVersion") ?? "auto"
        let sessionExpiryInterval = call.getInt("sessionExpiryInterval")
        let caFile = call.getString("caFile")
        let caPath = call.getString("caPath")
        
        let protocolVersion: MQTTClient.ProtocolVersion
        switch protocolVersionStr {
        case "5.0": protocolVersion = .v5
        case "3.1.1": protocolVersion = .v311
        default: protocolVersion = .auto
        }

        guard !host.isEmpty, !clientId.isEmpty else {
            call.reject("host and clientId are required")
            return
        }

        Task {
            do {
                if let caFile = caFile {
                    setenv("MQTT_QUIC_CA_FILE", caFile, 1)
                } else if let bundled = bundledCaPath() {
                    setenv("MQTT_QUIC_CA_FILE", bundled, 1)
                } else {
                    unsetenv("MQTT_QUIC_CA_FILE")
                }
                if let caPath = caPath {
                    setenv("MQTT_QUIC_CA_PATH", caPath, 1)
                } else {
                    unsetenv("MQTT_QUIC_CA_PATH")
                }
                if !NGTCP2Client.ping(host: host, port: UInt16(port)) {
                    DispatchQueue.main.async { call.reject("Server unreachable (UDP ping to \(host):\(port) failed). Check network and firewall.") }
                    return
                }
                // Idempotent / prevent concurrent connect: avoid second call disconnecting or replacing client (server sees stream reset)
                switch client.getState() {
                case .connected:
                    DispatchQueue.main.async {
                        call.resolve(["connected": true])
                        self.notifyListeners("connected", data: ["connected": true])
                    }
                    return
                case .connecting:
                    DispatchQueue.main.async { call.reject("Already connecting") }
                    return
                case .disconnected, .error:
                    break
                }
                client = MQTTClient(protocolVersion: protocolVersion)
                // Forward every incoming PUBLISH to JS so addListener('message', ...) receives topic + payload (matches Android)
                client.onPublish = { [weak self] topic, payload in
                    guard let self = self else { return }
                    let payloadStr: String = {
                        if let str = String(data: payload, encoding: .utf8) { return str }
                        return payload.base64EncodedString()
                    }()
                    DispatchQueue.main.async {
                        self.notifyListeners("message", data: ["topic": topic as String, "payload": payloadStr as String])
                    }
                }
                try await client.connect(
                    host: host,
                    port: UInt16(port),
                    clientId: clientId,
                    username: username,
                    password: password,
                    cleanSession: cleanSession,
                    keepalive: UInt16(keepalive),
                    sessionExpiryInterval: sessionExpiryInterval != nil ? UInt32(sessionExpiryInterval!) : nil
                )
                DispatchQueue.main.async {
                    call.resolve(["connected": true])
                    self.notifyListeners("connected", data: ["connected": true])
                }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }

    @objc func testHarness(_ call: CAPPluginCall) {
        let host = call.getString("host") ?? ""
        let port = call.getInt("port") ?? 1884
        let clientId = call.getString("clientId") ?? "AcharyaAnnadata"
        let topic = call.getString("topic") ?? "test/topic"
        let payload = call.getString("payload") ?? "Hello QUIC!"
        let caFile = call.getString("caFile")
        let caPath = call.getString("caPath")

        if host.isEmpty {
            call.reject("host is required")
            return
        }

        Task {
            do {
                if let caFile = caFile {
                    setenv("MQTT_QUIC_CA_FILE", caFile, 1)
                } else if let bundled = bundledCaPath() {
                    setenv("MQTT_QUIC_CA_FILE", bundled, 1)
                } else {
                    unsetenv("MQTT_QUIC_CA_FILE")
                }
                if let caPath = caPath {
                    setenv("MQTT_QUIC_CA_PATH", caPath, 1)
                } else {
                    unsetenv("MQTT_QUIC_CA_PATH")
                }

                client = MQTTClient(protocolVersion: .auto)
                try await client.connect(
                    host: host,
                    port: UInt16(port),
                    clientId: clientId,
                    username: nil,
                    password: nil,
                    cleanSession: true,
                    keepalive: 20,
                    sessionExpiryInterval: nil
                )
                try await client.subscribe(topic: topic, qos: 0, subscriptionIdentifier: nil)
                try await client.publish(topic: topic, payload: Data(payload.utf8), qos: 0)
                try await client.disconnect()
                DispatchQueue.main.async {
                    call.resolve(["success": true])
                    self.notifyListeners("subscribed", data: ["topic": topic])
                }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        Task {
            do {
                try await client.disconnect()
                DispatchQueue.main.async { call.resolve() }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }

    @objc func publish(_ call: CAPPluginCall) {
        let topic = call.getString("topic") ?? ""
        let qos = call.getInt("qos") ?? 0
        let messageExpiryInterval = call.getInt("messageExpiryInterval")
        let contentType = call.getString("contentType")
        
        guard !topic.isEmpty else {
            call.reject("topic is required")
            return
        }

        let data: Data
        if let payloadStr = call.getString("payload") {
            data = Data(payloadStr.utf8)
        } else if let arr = call.getArray("payload") as? [NSNumber] {
            data = Data(arr.map { UInt8(truncating: $0) })
        } else {
            call.reject("payload must be string or number array")
            return
        }

        Task {
            do {
                var properties: [UInt8: Any]? = nil
                if messageExpiryInterval != nil || contentType != nil {
                    properties = [:]
                    if let mei = messageExpiryInterval {
                        properties![MQTT5PropertyType.messageExpiryInterval.rawValue] = UInt32(mei)
                    }
                    if let ct = contentType {
                        properties![MQTT5PropertyType.contentType.rawValue] = ct
                    }
                }
                try await client.publish(topic: topic, payload: data, qos: UInt8(min(qos, 2)), properties: properties)
                DispatchQueue.main.async { call.resolve(["success": true]) }
            } catch {
                let msg = "\(error)"
                let code = (msg.contains("not connected") || msg.lowercased().contains("stream") || msg.lowercased().contains("connection")) ? "CONNECTION_LOST" : "PUBLISH_FAILED"
                DispatchQueue.main.async { call.reject(msg, code, nil) }
            }
        }
    }

    @objc func subscribe(_ call: CAPPluginCall) {
        let topic = call.getString("topic") ?? ""
        let qos = call.getInt("qos") ?? 0
        let subscriptionIdentifier = call.getInt("subscriptionIdentifier")

        guard !topic.isEmpty else {
            call.reject("topic is required")
            return
        }

        Task {
            do {
                try await client.subscribe(topic: topic, qos: UInt8(min(qos, 2)), subscriptionIdentifier: subscriptionIdentifier)
                // Incoming PUBLISH delivered via onPublish set in connect(); no per-topic handler needed
                DispatchQueue.main.async { call.resolve(["success": true]) }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }

    @objc func unsubscribe(_ call: CAPPluginCall) {
        let topic = call.getString("topic") ?? ""

        guard !topic.isEmpty else {
            call.reject("topic is required")
            return
        }

        Task {
            do {
                try await client.unsubscribe(topic: topic)
                DispatchQueue.main.async { call.resolve(["success": true]) }
            } catch {
                DispatchQueue.main.async { call.reject("\(error)") }
            }
        }
    }
}

File: android/src/main/kotlin/ai/annadata/mqttquic/MqttQuicPlugin.kt
package ai.annadata.mqttquic

import ai.annadata.mqttquic.client.MQTTClient
import ai.annadata.mqttquic.mqtt.MQTT5PropertyType
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import android.os.Handler
import android.os.Looper
import android.system.Os
import android.util.Base64
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.nio.charset.StandardCharsets
import java.io.File
import java.io.IOException
import java.net.InetAddress

/**
 * Capacitor plugin bridge. Phase 3: connect/publish/subscribe call into MQTTClient.
 */
@CapacitorPlugin(name = "MqttQuic")
class MqttQuicPlugin : Plugin() {

    private var client = MQTTClient(MQTTClient.ProtocolVersion.AUTO)
    private val scope = CoroutineScope(Dispatchers.Main)

    /** Last resolved IP per host (used when DNS fails on reconnect). */
    @Volatile
    private var lastResolvedHost: String? = null
    @Volatile
    private var lastResolvedIp: String? = null

    /** Resolve hostname to IP for socket connect (avoids "No address" on reconnect). Returns null if resolution fails. */
    private fun resolveHostToIp(host: String): String? {
        return try {
            InetAddress.getAllByName(host).firstOrNull()?.hostAddress?.also { ip ->
                lastResolvedHost = host
                lastResolvedIp = ip
            }
        } catch (e: Exception) {
            null
        }
    }

    /** Use cached IP for this host if fresh resolve failed (e.g. on reconnect). */
    private fun resolveOrCachedIp(host: String): String? {
        return resolveHostToIp(host) ?: if (host == lastResolvedHost) lastResolvedIp else null
    }

    private fun bundledCaFilePath(): String? {
        return try {
            val assetName = "mqttquic_ca.pem"
            val content = context.assets.open(assetName).bufferedReader().use { it.readText() }
            if (!content.contains("BEGIN CERTIFICATE")) {
                return null
            }
            val outFile = File(context.filesDir, assetName)
            outFile.writeText(content)
            outFile.absolutePath
        } catch (_: IOException) {
            null
        }
    }

    @PluginMethod
    fun connect(call: PluginCall) {
        val host = call.getString("host") ?: ""
        val port = call.getInt("port") ?: 1884
        val clientId = call.getString("clientId") ?: ""
        val username = call.getString("username")
        val password = call.getString("password")
        val cleanSession = call.getBoolean("cleanSession", true)
        val keepalive = call.getInt("keepalive", 20)
        val protocolVersionStr = call.getString("protocolVersion") ?: "auto"
        val sessionExpiryInterval = call.getInt("sessionExpiryInterval")
        val caFile = call.getString("caFile")
        val caPath = call.getString("caPath")
        
        val protocolVersion = when (protocolVersionStr) {
            "5.0" -> MQTTClient.ProtocolVersion.V5
            "3.1.1" -> MQTTClient.ProtocolVersion.V311
            else -> MQTTClient.ProtocolVersion.AUTO
        }

        if (host.isEmpty() || clientId.isEmpty()) {
            call.reject("host and clientId are required")
            return
        }

        scope.launch {
            try {
                try {
                    val bundled = bundledCaFilePath()
                    when {
                        caFile != null -> Os.setenv("MQTT_QUIC_CA_FILE", caFile, true)
                        bundled != null -> Os.setenv("MQTT_QUIC_CA_FILE", bundled, true)
                        else -> Os.setenv("MQTT_QUIC_CA_FILE", "", true)
                    }
                    if (caPath != null) {
                        Os.setenv("MQTT_QUIC_CA_PATH", caPath, true)
                    } else {
                        Os.setenv("MQTT_QUIC_CA_PATH", "", true)
                    }
                } catch (_: Exception) {
                    // Ignore env setup failures; native layer will report verification errors.
                }
                if (client.getState() == MQTTClient.State.CONNECTED) {
                    client.disconnect()
                }
                client = MQTTClient(protocolVersion)
                // Forward every incoming PUBLISH to JS so addListener('message', ...) receives topic + payload
                client.onPublish = { topic, payload ->
                    val payloadStr = try {
                        String(payload, StandardCharsets.UTF_8)
                    } catch (_: Exception) {
                        Base64.encodeToString(payload, Base64.NO_WRAP)
                    }
                    // Ensure non-null strings so Capacitor bridge never receives undefined
                    val safeTopic = topic ?: ""
                    val safePayload = payloadStr ?: ""
                    val data = JSObject().put("topic", safeTopic).put("payload", safePayload)
                    Handler(Looper.getMainLooper()).post {
                        notifyListeners("message", data)
                    }
                }
                // Resolve host to IP on IO so native getaddrinfo gets an IP (avoids "No address associated with hostname" on reconnect)
                val noAddressMsg = "No address associated with hostname"
                var lastException: Exception? = null
                for (attempt in 1..2) {
                    try {
                        withContext(Dispatchers.IO) {
                            val resolvedIp = resolveOrCachedIp(host)
                            client.connect(host, port, clientId, username, password, cleanSession ?: true, keepalive ?: 20, sessionExpiryInterval, connectAddress = resolvedIp)
                        }
                        // Cache resolved IP from native so reconnect can use it when Java DNS fails
                        client.getLastResolvedAddress()?.let { ip ->
                            lastResolvedHost = host
                            lastResolvedIp = ip
                        }
                        call.resolve(JSObject().put("connected", true))
                        notifyListeners("connected", JSObject().put("connected", true))
                        return@launch
                    } catch (e: Exception) {
                        // Cache resolved IP from native even on failure (e.g. CONNACK timeout) so reconnect can use it
                        client.getLastResolvedAddress()?.let { ip ->
                            lastResolvedHost = host
                            lastResolvedIp = ip
                        }
                        lastException = e
                        if (attempt == 1 && e.message?.contains(noAddressMsg, ignoreCase = true) == true) {
                            delay(2000L)
                            continue
                        }
                        break
                    }
                }
                call.reject(lastException?.message ?: "Connection failed")
            } catch (e: Exception) {
                call.reject(e.message ?: "Connection failed")
            }
        }
    }

    @PluginMethod
    fun testHarness(call: PluginCall) {
        val host = call.getString("host") ?: ""
        val port = call.getInt("port") ?: 1884
        val clientId = call.getString("clientId") ?: "mqttquic_test_client"
        val topic = call.getString("topic") ?: "test/topic"
        val payload = call.getString("payload") ?: "Hello QUIC!"
        val caFile = call.getString("caFile")
        val caPath = call.getString("caPath")

        if (host.isEmpty()) {
            call.reject("host is required")
            return
        }

        scope.launch {
            try {
                try {
                    val bundled = bundledCaFilePath()
                    when {
                        caFile != null -> Os.setenv("MQTT_QUIC_CA_FILE", caFile, true)
                        bundled != null -> Os.setenv("MQTT_QUIC_CA_FILE", bundled, true)
                        else -> Os.setenv("MQTT_QUIC_CA_FILE", "", true)
                    }
                    if (caPath != null) {
                        Os.setenv("MQTT_QUIC_CA_PATH", caPath, true)
                    } else {
                        Os.setenv("MQTT_QUIC_CA_PATH", "", true)
                    }
                } catch (_: Exception) {
                    // Ignore env setup failures; native layer will report verification errors.
                }

                client = MQTTClient(MQTTClient.ProtocolVersion.AUTO)
                client.connect(host, port, clientId, null, null, true, 60, null)
                client.subscribe(topic, 0, null)
                client.publish(topic, payload.toByteArray(StandardCharsets.UTF_8), 0, null)
                client.disconnect()
                call.resolve(JSObject().put("success", true))
                notifyListeners("subscribed", JSObject().put("topic", topic))
            } catch (e: Exception) {
                call.reject(e.message ?: "Test harness failed")
            }
        }
    }

    /**
     * Send MQTT PINGREQ and wait for PINGRESP. Resets server's idle timer.
     * Returns { ok: true } if PINGRESP received within timeout, { ok: false } on timeout or error.
     */
    @PluginMethod
    fun sendKeepalive(call: PluginCall) {
        val timeoutMs = call.getInt("timeoutMs", 5000)
        scope.launch {
            try {
                val ok = client.sendMqttPing((timeoutMs ?: 5000).toLong().coerceIn(1000, 15000))
                call.resolve(JSObject().put("ok", ok))
            } catch (e: Exception) {
                call.reject(e.message ?: "sendKeepalive failed")
            }
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        scope.launch {
            try {
                client.disconnect()
                call.resolve()
            } catch (e: Exception) {
                call.reject(e.message ?: "Disconnect failed")
            }
        }
    }

    @PluginMethod
    fun publish(call: PluginCall) {
        val topic = call.getString("topic") ?: ""
        val qos = call.getInt("qos", 0)
        val messageExpiryInterval = call.getInt("messageExpiryInterval")
        val contentType = call.getString("contentType")

        if (topic.isEmpty()) {
            call.reject("topic is required")
            return
        }

        val data: ByteArray = when {
            call.getString("payload") != null ->
                call.getString("payload")!!.toByteArray(StandardCharsets.UTF_8)
            call.getArray("payload") != null -> {
                val arr = call.getArray("payload")!!
                (0 until arr.length()).mapNotNull { i ->
                    (arr.get(i) as? Number)?.toInt()?.and(0xFF)?.toByte()
                }.toByteArray()
            }
            else -> {
                call.reject("payload must be string or number array")
                return
            }
        }

        scope.launch {
            try {
                val properties = mutableMapOf<Int, Any>()
                messageExpiryInterval?.let { properties[MQTT5PropertyType.MESSAGE_EXPIRY_INTERVAL.toInt()] = it }
                contentType?.let { properties[MQTT5PropertyType.CONTENT_TYPE.toInt()] = it }
                client.publish(topic, data, minOf(qos ?: 0, 2), if (properties.isNotEmpty()) properties else null)
                call.resolve(JSObject().put("success", true))
            } catch (e: Exception) {
                val msg = e.message ?: "Publish failed"
                val code = when {
                    msg.contains("not connected", ignoreCase = true) -> "CONNECTION_LOST"
                    msg.contains("stream", ignoreCase = true) || msg.contains("connection", ignoreCase = true) -> "CONNECTION_LOST"
                    else -> "PUBLISH_FAILED"
                }
                call.reject(msg, code)
            }
        }
    }

    @PluginMethod
    fun subscribe(call: PluginCall) {
        val topic = call.getString("topic") ?: ""
        val qos = call.getInt("qos", 0)
        val subscriptionIdentifier = call.getInt("subscriptionIdentifier")

        if (topic.isEmpty()) {
            call.reject("topic is required")
            return
        }

        scope.launch {
            try {
                client.subscribe(topic, minOf(qos ?: 0, 2), subscriptionIdentifier)
                call.resolve(JSObject().put("success", true))
            } catch (e: Exception) {
                call.reject(e.message ?: "Subscribe failed")
            }
        }
    }

    @PluginMethod
    fun unsubscribe(call: PluginCall) {
        val topic = call.getString("topic") ?: ""

        if (topic.isEmpty()) {
            call.reject("topic is required")
            return
        }

        scope.launch {
            try {
                client.unsubscribe(topic)
                call.resolve(JSObject().put("success", true))
            } catch (e: Exception) {
                call.reject(e.message ?: "Unsubscribe failed")
            }
        }
    }
}
